webapp/sources/rudder/rudder-core/src/main/scala/com/normation/plugins/PluginSettings.scala

```diff
@@ -93,73 +94,153 @@ final case class JsonPluginsDetails(
+  import GlobalPluginsLicense.*
+
+  // for efficiency : check equality and hash first before field comparison,
+  // as it will mostly be the case because license information should be similar
+  def combine(that: GlobalPluginsLicense[EndDate])(implicit endSemigroup: Semigroup[EndDate]) = {
+    def combineOptField[A](toa: GlobalPluginsLicense[EndDate] => Option[A])(f: (A, A) => A): Option[A] = {
+      implicit val s: Semigroup[A] = Semigroup.instance(f)
+      Semigroup[Option[A]].combine(toa(this), toa(that))
+    }
     if (this == that) this
     else {
-      JsonGlobalPluginLimits(
-        comp[NonEmptyChunk[String]](this.licensees, that.licensees, _ ++ _),
-        comp[ZonedDateTime](this.startDate, that.startDate, (x, y) => if (x.isAfter(y)) x else y),
-        comp[ZonedDateTime](this.endDate, that.endDate, (x, y) => if (x.isBefore(y)) x else y),
-        comp[Int](this.maxNodes, that.maxNodes, (x, y) => if (x < y) x else y)
-      )
+      new GlobalPluginsLicense[EndDate](
+        combineOptField(_.licensees)(_ ++ _),
+        combineOptField(_.startDate)((x, y) => if (x.isAfter(y)) x else y),
+        combineOptField(_.endDate)(endSemigroup.combine),
+        combineOptField(_.maxNodes)(_.min(_))
+      ) {}
     }
   }
+
+  private[GlobalPluginsLicense] def sortDistinctLicensees: GlobalPluginsLicense[EndDate] = {
+    withLicensees(licensees.map(_.sorted.distinct).flatMap(NonEmptyChunk.fromChunk))
+  }
+
+  protected def withLicensees(licensees: Option[NonEmptyChunk[String]]): GlobalPluginsLicense[EndDate] = {
+    new GlobalPluginsLicense[EndDate](
+      licensees,
+      startDate,
+      endDate,
+      maxNodes
+    ) {}
+  }
 }
-object JsonGlobalPluginLimits {
+
+object GlobalPluginsLicense {
   import DateFormaterService.JodaTimeToJava
-  import DateFormaterService.json.encoderZonedDateTime
-  implicit val encoderGlobalPluginLimits: JsonEncoder[JsonGlobalPluginLimits] = DeriveJsonEncoder.gen
 
-  def fromLicenseInfo(info: PluginLicenseInfo): JsonGlobalPluginLimits = {
-    JsonGlobalPluginLimits(
+  /**
+    * Typeclass for proving that a type can be constructed from a Java ZonedDateTime.
+    * This should be a functor to allow building wrapping datastructures but is limited to supported ones for now.
+    */
+  sealed private trait ToEndDate[T] {
+    def fromZonedDateTime(date: ZonedDateTime): T
+  }
+  private object ToEndDate          {
+    def apply[T](implicit ev: ToEndDate[T]) = ev
+
+    implicit val id: ToEndDate[ZonedDateTime] = new ToEndDate[ZonedDateTime] {
+      def fromZonedDateTime(date: ZonedDateTime): ZonedDateTime = date
+    }
+
+    implicit val dateCounts: ToEndDate[DateCounts] = new ToEndDate[DateCounts] {
+      override def fromZonedDateTime(date: ZonedDateTime): DateCounts = DateCounts.one(date)
+    }
+  }
+
+  final case class DateCount(date: ZonedDateTime, count: Int)
+  // Opaque type for counts by date. Encoded as json list but using a Map for unicity when grouping by date
```

Comment:
that's not opaque at all ;)
Can you use "dedicated data structure" until we really switch to scala 3 and are able to use opaque-type-in-the-scala-meaning to avoid confusion ?